### PR TITLE
[http2] Fix Http2Adapter not redirecting for relative URLs

### PR DESCRIPTION
**Pull request by kuhnroyal**
Fixes a redirect bug discovered during https://github.com/cfug/dio/pull/2077 where the adapter would fail to properly redirect if the location header contains a relative URL.

### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### CODE REVIEW

File: plugins/http2_adapter/lib/src/http2_adapter.dart [hunk 1]
```diff
@@ -196,11 +196,34 @@ class Http2Adapter implements HttpClientAdapter {
     // Handle redirection.
     if (needRedirect) {
       final url = responseHeaders.value('location');
+      if (url == null) {
```
Comment (AlexV525) [inline plugins/http2_adapter/lib/src/http2_adapter.dart:199]: Do we have a RFC for the behavior? I can come out with another idea that we might hanfle this as a self-redirection. Someone else could thought the same thing.

File: plugins/http2_adapter/test/http2_test.dart
Comment (AlexV525) [inline plugins/http2_adapter/test/http2_test.dart:1]: Can we also test the empty location?

File: plugins/http2_adapter/lib/src/http2_adapter.dart (hunk 1 above)
Comment (kuhnroyal) [reply plugins/http2_adapter/lib/src/http2_adapter.dart:199]: No, that just didn't make sense to me. Needs to be changed back.

File: plugins/http2_adapter/test/http2_test.dart
Comment (kuhnroyal) [reply plugins/http2_adapter/test/http2_test.dart:1]: I'll try, probably not against httpbun.

Comment (kuhnroyal) [reply plugins/http2_adapter/test/http2_test.dart:1]: No way to test this against httpbun. Might have to extract this into some helper function to make it testable.

**Comment by AlexV525** [comment]
@kuhnroyal Still WIP?

**Comment by kuhnroyal** [comment]
Updated

File: plugins/http2_adapter/test/http2_test.dart
Comment (kuhnroyal) [reply plugins/http2_adapter/test/http2_test.dart:1]: Added a util method and separate tests since empty location header is not supported by httpbun.

File: plugins/http2_adapter/lib/src/http2_adapter.dart (hunk 1 above)
Comment (kuhnroyal) [reply plugins/http2_adapter/lib/src/http2_adapter.dart]: Fixed

**Review by AlexV525 — APPROVED** [review]
👍
